### PR TITLE
Reduce memory allocation during address change notification.

### PR DESCRIPTION
**Pull request by chickenlj**
1. Merge registry notification events happened in a certain short interval.
2. cache URL instances to reduce string -> URL recreation
3. reduce allocation during export and refer at start up.
4. add frozen status for URL
5. use URLBuilder for unfrozen URLs

### CODE REVIEW

**Comment by guohao** [comment]
👍

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 1]
```diff
@@ -203,20 +204,17 @@ public URL(String protocol,
...
-        this.address = getAddress(this.host, this.port);
 
         // trim the beginning "/"
         while (path != null && path.startsWith("/")) {
             path = path.substring(1);
         }
         this.path = path;
         if (parameters == null) {
-            parameters = new HashMap<>();
+            this.parameters = new HashMap<>();
```
Comment (guohao) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java]: `Collections.emptyMap()` will be better ?

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/RegistryNotifier.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        public NotificationTask(RegistryNotifier listener, long time) {
+            this.listener = listener;
+            this.time = time;
+        }
+
+        @Override
+        public void run() {
+            int delayTime = getRegistry().getDelay();
+            if (this.time < listener.lastEventTime) {
```
Comment (guohao) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/RegistryNotifier.java]: Corner case:  Continuing notification from registry may cause the listener nerver  be notified.(`lastEventTime` is always greater than `this.time`). 
At most one running `listener.doNotifyTask` task will  be better :P

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/RegistryNotifier.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                return;
+            } else if (delayTime <= 0) {
+                listener.doNotify(listener.rawAddresses);
+                listener.lastExecuteTime = System.currentTimeMillis();
+            } else if (System.currentTimeMillis() - listener.lastExecuteTime < delayTime) {
+                scheduler.schedule(new NotificationTask(this.listener, this.time), System.currentTimeMillis() - listener.lastExecuteTime, TimeUnit.MILLISECONDS);
```
Comment (guohao) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/RegistryNotifier.java]: `Math.max(0,delayTime-(System.currentTimeMillis() - listener.lastExecuteTime))`

File: dubbo-registry/dubbo-registry-etcd3/src/main/java/org/apache/dubbo/registry/etcd/EtcdRegistry.java [hunk 4]
```diff
@@ -196,6 +196,7 @@ public void doSubscribe(URL url, NotifyListener listener) {
                             CHECK_KEY, String.valueOf(false)), listener);
                 }
             } else {
+                CountDownLatch latch = new CountDownLatch(1);
```
Comment (guohao) [inline dubbo-registry/dubbo-registry-etcd3/src/main/java/org/apache/dubbo/registry/etcd/EtcdRegistry.java]: unused

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java [hunk 5]
```diff
@@ -50,71 +41,42 @@ private ClusterUtils() {
...
-                copyOfLocalMap.remove(VERSION_KEY);
-            }
+        // Remove configurations from provider, some items should not being affected by provider.
+        remoteMap.remove(THREAD_NAME_KEY);
+        remoteMap.remove(THREADPOOL_KEY);
+        remoteMap.remove(CORE_THREADS_KEY);
+        remoteMap.remove(THREADS_KEY);
+        remoteMap.remove(QUEUES_KEY);
+        remoteMap.remove(ALIVE_KEY);
+        remoteMap.remove(TRANSPORTER_KEY);
```
Comment (LinShunKang) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java]: if `localMap` is empty, `remoteMap` will loss these entries, I'm not sure if this works as expected.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java [hunk 6]
```diff
@@ -50,71 +41,42 @@ private ClusterUtils() {
...
-
-            map.remove(QUEUES_KEY);
-            map.remove(DEFAULT_KEY_PREFIX + QUEUES_KEY);
-
-            map.remove(ALIVE_KEY);
-            map.remove(DEFAULT_KEY_PREFIX + ALIVE_KEY);
-
-            map.remove(Constants.TRANSPORTER_KEY);
-            map.remove(DEFAULT_KEY_PREFIX + Constants.TRANSPORTER_KEY);
+        if (remoteMap == null || remoteMap.size() <= 0) {
```
Comment (LinShunKang) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java]: Use `MapUtils.isEmpty(remoteMap)` is better ?

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java [hunk 7]
```diff
@@ -50,71 +41,42 @@ private ClusterUtils() {
...
+            String remoteApplication = remoteMap.get(APPLICATION_KEY);
+            remoteMap.putAll(localMap);
+            remoteMap.put(REMOTE_APPLICATION_KEY, remoteApplication);
 
             // Combine filters and listeners on Provider and Consumer
             String remoteFilter = remoteMap.get(REFERENCE_FILTER_KEY);
-            String localFilter = copyOfLocalMap.get(REFERENCE_FILTER_KEY);
+            String localFilter = localMap.get(REFERENCE_FILTER_KEY);
             if (remoteFilter != null && remoteFilter.length() > 0
                     && localFilter != null && localFilter.length() > 0) {
```
Comment (LinShunKang) [inline dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java]: Use `StringUtils.isNotEmpty(remoteFilter) && StringUtils.isNotEmpty(localFilter)` is better ?

File: dubbo-common/src/main/java/org/apache/dubbo/common/URL.java [hunk 8]
```diff
@@ -1139,18 +1143,34 @@ public URL addParameters(Map<String, String> parameters) {
             return this;
         }
 
-        Map<String, String> map = new HashMap<>(getParameters());
-        map.putAll(parameters);
-        return new URL(protocol, username, password, host, port, path, map);
+        if (frozen) {
+            Map<String, String> map = new HashMap<>(getParameters());
+            map.putAll(parameters);
```
Comment (LinShunKang) [inline dubbo-common/src/main/java/org/apache/dubbo/common/URL.java]: Create map with expect size would reduce memory, just like this: 
```
Map<String, String> srcParams = getParameters();
Map<String, String> newMap = new HashMap<>((int)((srcParams.size() + parameters.size) / 0.75 + 1));
newMap.putAll(srcParams);
newMap.putAll(parameters);
```

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/RegistryNotifier.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+import org.apache.dubbo.common.extension.ExtensionLoader;
+import org.apache.dubbo.common.threadpool.manager.ExecutorRepository;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public abstract class RegistryNotifier {
+    private long lastExecuteTime;
+    private long lastEventTime;
```
[the remark this answers is not in the document]
Comment (chickenlj) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/RegistryNotifier.java]: For most registries, It's a single-threaded model.  But I think we can add volatile.

File: dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java (hunk 5 above)
Comment (chickenlj) [reply dubbo-cluster/src/main/java/org/apache/dubbo/rpc/cluster/support/ClusterUtils.java]: Yes, that is what we want, these entries only work for the provider side.

**Comment by hengyunabc** [comment]
这个PR被 revert掉了，原因是：

原来的url：

```
dubbo://127.0.0.1:9098/org.apache.dubbo.registry.dubbo.RegistryDirectoryTest$DemoService?check=false&interface=org.apache.dubbo.registry.dubbo.RegistryDirectoryTest$DemoService&lazy=true&methods=getXXX&remote.application=
```

精简后的url:

```
dubbo://127.0.0.1:9098?check=false&lazy=true&methods=getXXX&refer=interface%3Dorg.apache.dubbo.registry.dubbo.RegistryDirectoryTest%24DemoService&remote.application=
```

1. 精简过后的url里没有了 path
2. 会导致代码里原来用 url.getPath() 来取得 service name的逻辑失效
3. 如果客户依赖这个行为，就会出现不兼容问题
4. 需要再仔细评估

**Pull request merged**